hack/generate/index.sh

```diff
@@ -2,29 +2,85 @@
-values[LATEST_VERSIONED_CHANNEL]="$(metadata.get 'olm.channels.list[*]' | head -n 2 | tail -n 1)"
-values[PREVIOUS_CHANNEL]="$(metadata.get 'olm.channels.list[*]' | head -n 3 | tail -n 1)"
-values[PREVIOUS_REPLACES_CHANNEL]="$(metadata.get 'olm.channels.list[*]' | head -n 4 | tail -n 1)"
+  # Initialize a temporary file for a single entry.
+  cat > "${channel_entry_yaml}" <<EOF
+schema: olm.channel
+name: ${channel}
+package: serverless-operator
+entries:
+EOF
 
-values[PREVIOUS_CHANNEL_HEAD]="${values[PREVIOUS_CHANNEL]#stable-}.0"
-values[PREVIOUS_REPLACES_CHANNEL_HEAD]="${values[PREVIOUS_REPLACES_CHANNEL]#stable-}.0"
+  current_version=$(metadata.get 'project.version')
+  major=$(versions.major "$current_version")
+  minor=$(versions.minor "$current_version")
+  micro=$(versions.micro "$current_version")
 
-# Start fresh
-cp "$template" "$target"
+  # Handle the first entry specifically as it might be a z-stream release.
+  if [[ "$micro" == "0" ]]; then
+    previous_version="${major}.$(( minor-1 )).${micro}"
+  else
+    previous_version="${major}.${minor}.0"
+  fi
 
-for before in "${!values[@]}"; do
-  echo "Value: ${before} -> ${values[$before]}"
-  sed --in-place "s/__${before}__/${values[${before}]}/" "$target"
-done
+  cat << EOF | yq write --inplace --script - "$channel_entry_yaml"
+  - command: update
+    path: entries[+]
+    value:
+      name: "serverless-operator.v${current_version}"
+      replaces: "serverless-operator.v${previous_version}"
+      skipRange: ">=${previous_version} <${current_version}"
+EOF
+
+  # One is already added above specifically
+  num_csvs=$(( INDEX_IMAGE_NUM_CSVS-1 ))
+
+  # Generate additional entries
+  for i in $(seq $num_csvs); do
+    current_minor=$(( minor-i ))
+    previous_minor=$(( minor-i ))
+    previous_minor=$(( previous_minor-1 ))
+    # If the current version is a z-stream then the following entries will
+    # start with the same "minor" version.
+    if [[ "$micro" != "0" ]]; then
+      current_minor=$(( current_minor+1 ))
+      previous_minor=$(( previous_minor+1 ))
+    fi
+
+    current_version="${major}.${current_minor}.0"
+    previous_version="${major}.${previous_minor}.0"
+
+    # If this is the last item enter only name, without "replaces".
+    if [[ $i -eq $num_csvs ]]; then
+      yq write --inplace "$channel_entry_yaml" 'entries[+].name' "serverless-operator.v${current_version}"
+    else
+      cat << EOF | yq write --inplace --script - "$channel_entry_yaml"
+  - command: update
+    path: entries[+]
+    value:
+      name: "serverless-operator.v${current_version}"
+      replaces: "serverless-operator.v${previous_version}"
+      skipRange: ">=${previous_version} <${current_version}"
+EOF
+    fi
+  done
+
+  echo "---" >> "${target}"
+  cat "${channel_entry_yaml}" >> "${target}"
+}
+
+# Clear the file.
+rm -f "${target}"
+
+add_channel_entries "stable" "${target}"
+add_channel_entries "stable-1.34" "${target}"
```

Comment:
Should this values be taken from `project.yaml` ?

Reply:
Right. Good idea.